src/util/submit-addon.js

```diff
@@ -308,15 +313,80 @@ export default class Client {
     return promisify(pipeline)(contents, createWriteStream(destPath));
   }
 
-  async postNewAddon(
+  /*
+  This function aims to quickly hash the contents of the zip file that's being uploaded,
+  to compare it to the previous zip file that was uploaded, so we can skip the upload for
+  efficiency.
+
+  CRCs are used from the zip to avoid having to extract and hash  all the files.
+
+  Two zips that have different byte contents in their files must have a different hash;
+  but returning a different hash when the contents are the same in some cases is acceptable
+  - a false mismatch does not result in lost data.
+  */
+  async hashXpiCrcs(
+    filePath: string,
+    asyncFsReadFile: typeof defaultAsyncFsReadFile = defaultAsyncFsReadFile
+  ): Promise<string> {
+    const zip = await JSZip.loadAsync(
+      asyncFsReadFile(filePath, { createFolders: true })
+    );
+    const hash = createHash('sha256');
+    const entries = [];
+    zip.forEach((relativePath, entry) => {
+      let path = relativePath;
+      if (entry.dir) {
+        path += '/';
+      }
+      entries.push({ path, crc32: entry._data?.crc32 || 0 });
```

Comment:
Would be nice to add a comment explaining why this is correct, even if it is just a link to my previous comment.

Reply:
added